### DISCUSSION TITLE
Method that Defines and Returns Class Instance

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [python/typing#1249] [Q&A]
I am using a third-party library that defines a method like so:

```python
# some_module.py

def func(cls_to_wrap):
    class A(cls_to_wrap):
        ...
    return A()
```

The function, class, and class methods are all not type-hinted, so the above is an accurate representation. My problem is when using this method:

```
wrapped = func(some_class)
```

`mypy` rightly is unable to determine that `wrapped` has any attributes defined by `A` and gives an `attr-defined` error. I cannot use something like `assert isinstance(wrapped, A)` because `A` is not exposed in `some_module.py`.

Being that this is a third-party library that I cannot change, is my only option at this point to use `# typing: ignore` to disable the warning message?

**Comment by TeamSpen210** [comment]
You could declare it as `def func(cls: Type[T]) -> T:`, but that doesn't include the additional attributes `A` offers. What you need is `-> Intersection[T, ProtocolOfA]` (#213), where you define a (possibly generic) protocol outside the function with the methods contained. But we don't have intersections. Perhaps you could do `def func(cls: Type[ProtocolA[T1, T2]]) -> ProtocolB[T1, T2]:`, with one to match the parent class and extract the defined types, then a second to declare the combination...

**Reply by srittau** [reply]
You could define your own `@runtime_checkable` protocol with the fields of `A` and use `isinstance` against that.

```python
from typing import runtime_checkable, Protocol

@runtime_checkable
class A(Protocol):
   ...

wrapped = func(some_class)
assert isinstance(wrapped, A)
```
